smart_card_connector_app/src/testing_smart_card_simulation.cc

```diff
@@ -77,10 +322,322 @@ void TestingSmartCardSimulation::PostFakeJsReply(RequestId request_id,
+TestingSmartCardSimulation::FindDeviceStateByIdAndHandleLocked(
+    int64_t device_id,
+    int64_t device_handle) {
+  DeviceState* device_state = FindDeviceStateByIdLocked(device_id);
+  if (device_state && device_state->opened_device_handle == device_handle)
+    return device_state;
+  return nullptr;
+}
+
 void TestingSmartCardSimulation::OnListDevicesCalled(RequestId request_id) {
-  // Pretend to have an empty list of devices.
-  PostFakeJsReply(request_id,
-                  ArrayValueBuilder().Add(Value(Value::Type::kArray)).Get());
+  std::vector<std::unique_ptr<Value>> libusb_js_devices;
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    for (const auto& device_state : device_states_) {
+      libusb_js_devices.push_back(MakeUnique<Value>(
+          ConvertToValueOrDie(MakeLibusbJsDevice(device_state.device))));
+    }
+  }
+
+  // Reply outside mutex, to prevent potential deadlocks in case the libusb
+  // caller makes another request in the callback.
+  PostFakeJsResult(request_id, Value(std::move(libusb_js_devices)));
+}
+
+void TestingSmartCardSimulation::OnGetConfigurationsCalled(
+    int64_t device_id,
+    RequestId request_id) {
+  optional<DeviceType> device_type;
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    DeviceState* device_state = FindDeviceStateByIdLocked(device_id);
+    if (device_state)
+      device_type = device_state->device.type;
+  }
+
+  // Reply outside mutex, to prevent potential deadlocks in case the libusb
+  // caller makes another request in the callback.
+  if (!device_type) {
+    PostFakeJsError(request_id, "Unknown device");
+    return;
+  }
+  PostFakeJsResult(
+      request_id,
+      ConvertToValueOrDie(MakeLibusbJsConfigurationDescriptors(*device_type)));
+}
+
+void TestingSmartCardSimulation::OnOpenDeviceHandleCalled(
+    int64_t device_id,
+    RequestId request_id) {
+  optional<int> handle;
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    DeviceState* device_state = FindDeviceStateByIdLocked(device_id);
+    if (device_state) {
+      GOOGLE_SMART_CARD_CHECK(!device_state->opened_device_handle);
+      device_state->opened_device_handle = next_free_device_handle_;
+      handle = next_free_device_handle_;
+      ++next_free_device_handle_;
+    }
+  }
+
+  // Reply outside mutex, to prevent potential deadlocks in case the libusb
+  // caller makes another request in the callback.
+  if (!handle) {
+    PostFakeJsError(request_id, "Unknown device");
+    return;
+  }
+  PostFakeJsResult(request_id, Value(*handle));
+}
+
+void TestingSmartCardSimulation::OnCloseDeviceHandleCalled(
+    int64_t device_id,
+    int64_t device_handle,
+    RequestId request_id) {
+  bool device_found = false;
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    DeviceState* device_state =
+        FindDeviceStateByIdAndHandleLocked(device_id, device_handle);
+    if (device_state) {
+      device_found = true;
+      device_state->opened_device_handle = {};
```

Comment:
self-nit: Probably want to assert `device_state->opened_device_handle` is non-null.